packages/beacon-node/src/network/reqresp/utils/dataColumnResponseValidaiton.ts

```
@@ -0,0 +1,66 @@
+import {LogData} from "@lodestar/logger";
+import {ForkSeq, KZG_COMMITMENTS_GINDEX} from "@lodestar/params";
+import {RespStatus, ResponseError} from "@lodestar/reqresp";
+import {ColumnIndex, Slot} from "@lodestar/types";
+import {Logger, bytesToInt, toHex} from "@lodestar/utils";
+import {IBeaconChain} from "../../../chain/interface.js";
+import {IBeaconDb} from "../../../db/interface.js";
+
+export async function logDataColumnSidecarUnavailability(opts: {
+  chain: IBeaconChain;
+  db: IBeaconDb;
+  index: ColumnIndex;
+  slot: Slot;
+  blockRoot?: Uint8Array;
+  logData: Record<string, LogData>;
+}): Promise<void> {
+  const {chain, db, index, slot, blockRoot, logData} = opts;
+  let localLogData: LogData = {...logData, slot};
+
+  if (blockRoot) {
+    localLogData = {...localLogData, blockRoot: toHex(blockRoot)};
+  }
+
+  const blockBytes = blockRoot ? await db.block.getBinary(blockRoot) : await db.blockArchive.getBinary(slot);
+
+  if (!blockBytes) {
+    chain.logger.error("Expected unfinalized block not found", localLogData);
+    return;
+  }
+
+  const blobsCount =
+    chain.config.getForkSeq(slot) < ForkSeq.deneb
+      ? 0
+      : bytesToInt(blockBytes.slice(KZG_COMMITMENTS_GINDEX, KZG_COMMITMENTS_GINDEX + 4));
+
+  localLogData = {...localLogData, blobsCount, index};
+
+  if (blobsCount > 0) {
+    chain.logger.error("Requested dataColumnSidecar is missing.", localLogData);
```

in this case, I think we should throw `ResourceUnavailable` instead of just logging so that peers know we have an error in the response?

---

If we serve multiple requests throwing error will fail/abort all the requests. We may be able to serve the most of the dataColumn sidecars except one slot/block in that case it would be better to just log the error for us and not throw error. 

@matthewkeil Please share your thoughts.

---

Hmmm... I think @twoeths might be right.  Here are the pertinent sections of the spec.

https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/p2p-interface.md#datacolumnsidecarsbyrange-v1

```
Peers that are unable to reply to data column sidecar requests within the range
data_column_serve_range SHOULD respond with error code 3: ResourceUnavailable. Such 
peers that are unable to successfully reply to this range of requests MAY get descored or
disconnected at any time.

Clients MUST respond with at least the data column sidecars of the first blob-carrying block
that exists in the range, if they have it, and no more than 
MAX_REQUEST_DATA_COLUMN_SIDECARS sidecars.

Clients MUST include all data column sidecars of each block from which they include data 
column sidecars.

Slots that do not contain known data columns MUST be skipped, mimicking the behaviour of
the BlocksByRange request. Only response chunks with known data columns should therefore be sent.
```

---

After speak with the Prysm team it sounds like they are just ignoring data they do not have and returning what they do have.  Suggest we do the same and can come back and revisit later if issues arrise

---

Yes pls. Not only is it the simplest solution (just serve the data), its also what other teams are doing. (And it's reasonable to interpret the spec as this being the "proper" behavior)

---

it's really a bug if block has non-zero blob and we cannot serve custodied column, we should track this in metrics to investigate

> If we serve multiple requests throwing error will fail/abort all the requests. We may be able to serve the most of the dataColumn sidecars except one slot/block in that case it would be better to just log the error for us and not throw error.

I think the opposite, it's pretty hard to deal with that if we have a look at how we consume `beacon_data_column_sidecar_by_range` when we request data from peers. 

I prefer "all or nothing", if we response data it should be qualified data, it's tricky to response data and we know we miss some data in the response. Also lighthouse response error in this case [here](https://github.com/sigp/lighthouse/blob/cfb1f7331064b758c6786e4e1dc15507af5ff5d1/beacon_node/network/src/network_beacon_processor/rpc_methods.rs#L1091)

---

@matthewkeil looks like prysm also responses error in this case [here](https://github.com/OffchainLabs/prysm/blob/8ec460223c802439662428e1b931225fbb646bfb/beacon-chain/sync/rpc_data_column_sidecars_by_range.go#L145)

if you jump into `data_column.go` `Get` function there, it'll throw error if there is no DataColumnSidecar for an index